File: TinyGP/refactored 2/src/OutInterface.java
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;


public interface OutInterface {
    void print(String s);
    default void println(String s) {
        print(s + '\n');
    }

    default void print(Object o) {
        print(o.toString());
    }
    default void println(Object o) {
        print(o.toString() + '\n');
    }

    default void close() {}

    static OutInterface createOutputToFile(String filename) {
        try {
            File file = new File(filename);
            if (file.exists()) {
                System.out.println("File already exists: " + filename);
                System.out.print("Overwrite? [y/n] ");
                int input = System.in.read();
                System.in.read();
                if (input != 'y' && input != 'Y') {
                    System.out.println("Exiting...");
                    System.exit(0);
                }
                file.delete();
            }

            file.createNewFile();
            BufferedWriter writer = new BufferedWriter(new FileWriter(file));

            return new FileOutputInterface(writer);
        } catch (IOException e) {
            System.out.println("Error while creating output to file: " + filename);
            System.exit(1);
            return null;
        }
    }

    class FileOutputInterface implements OutInterface {
        private final BufferedWriter writer;

        public FileOutputInterface(BufferedWriter writer) {
            this.writer = writer;
        }

        @Override
        public void print(String s) {
            try {
                writer.write(s);
                writer.flush();
            } catch (IOException e) {
                System.out.println("Error while writing to file.");
                System.exit(1);
            }
        }
        
        @Override
        public void close() {
            try {
                writer.close();
            } catch (IOException e) {
                System.out.println("Error while closing file writer.");
                System.exit(1);
            }
        }
    }
}

File: OurGP/OutInterface.cs
namespace OutInterface;

public interface IOutInterface
{
    void Print(string s);
    void PrintLine(string s)
    {
        Print(s + "\n");
    }

    void Print() { }
    void PrintLine() { }
    
    void Print(object o)
    {
        Print(o.ToString()!);
    }
    void PrintLine(object o)
    {
        PrintLine(o.ToString()!);
    }
}

public class FileOutputInterface : IOutInterface
{
    private readonly StreamWriter writer;

    public FileOutputInterface(StreamWriter writer)
    {
        this.writer = writer;
    }

    public void Print(string s)
    {
        try
        {
            writer.Write(s);
            writer.Flush();
        }
        catch (IOException)
        {
            Console.WriteLine("Error while writing to file.");
            Environment.Exit(1);
        }
    }

    public void Close()
    {
        try
        {
            writer.Close();
        }
        catch (IOException)
        {
            Console.WriteLine("Error while closing file writer.");
            Environment.Exit(1);
        }
    }
}

public class ConsoleOutputInterface : IOutInterface
{
    public void Print(string s)
    {
        Console.Write(s);
    }
}

public static class OutInterfaceFactory
{
    public static IOutInterface CreateOutputToFile(string filename)
    {
        try
        {
            if (File.Exists(filename))
            {
                Console.WriteLine("File already exists: " + filename);
                Console.Write("Overwrite? [y/n] ");
                var input = Console.ReadLine();
                if (input != "y" && input != "Y")
                {
                    Console.WriteLine("Exiting...");
                    Environment.Exit(0);
                }
                File.Delete(filename);
            }

            var file = new StreamWriter(filename);
            return new FileOutputInterface(file);
        }
        catch (IOException)
        {
            Console.WriteLine("Error while creating output to file: " + filename);
            Environment.Exit(1);
            return null;
        }
    }

    public static IOutInterface CreateOutputToConsole()
    {
        return new ConsoleOutputInterface();
    }
}